src/main/java/org/springframework/up/git/UrlRepositoryService.java

```diff
@@ -0,0 +1,217 @@
+import org.springframework.up.config.TemplateRepositoryProperties;
+import org.springframework.util.FileSystemUtils;
+import org.springframework.util.ResourceUtils;
+import org.springframework.util.StreamUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author Thomas Risberg
+ */
+@Component
+public class UrlRepositoryService implements SourceRepositoryService {
+
+	private final Logger logger = LoggerFactory.getLogger(UrlRepositoryService.class);
+
+	private final TemplateRepositoryProperties templateRepositoryProperties;
+
+	public UrlRepositoryService(TemplateRepositoryProperties templateRepositoryProperties) {
+		this.templateRepositoryProperties = templateRepositoryProperties;
+	}
+
+	@Override
+	public Path retrieveRepositoryContents(String sourceRepoUrl) {
+		Path targetPath;
+		try {
+			targetPath = Files.createTempDirectory("source-repo-");
+		}
+		catch (IOException e) {
+			throw new UpException("Failed to create temp directory: " + e.getMessage(), e);
+		}
+
+		Path contentPath;
+		if (sourceRepoUrl.startsWith("file:")) {
+			contentPath = retrieveFileContents(sourceRepoUrl, targetPath);
+		}
+		else {
+			GitRepoUrlRef gitRepoUrlRef = GitRepoUrlRef.fromUriString(sourceRepoUrl);
+			if (gitRepoUrlRef.getRepoUrl().toString().contains("github.com")) {
+				contentPath = retrieveGitHubRepositoryContents(gitRepoUrlRef, targetPath);
+			}
+			else {
+				contentPath = retrieveGitLabRepositoryContents(gitRepoUrlRef, targetPath);
+			}
+		}
+		logger.debug("Source from " + sourceRepoUrl + " retrieved into " + contentPath.toFile().getAbsolutePath());
+		return contentPath;
+	}
+
+	/**
+	 * Retrieve contents from a file location.
+	 */
+	private Path retrieveFileContents(String source, Path targetPath) {
+		try {
+			File src = ResourceUtils.getFile(source);
+			File dest = targetPath.toFile();
+			logger.debug("Copying file resource: " + src + " to " + dest);
+			FileSystemUtils.copyRecursively(src, dest);
+		}
+		catch (IOException e) {
+			throw new UpException("Failed processing " + source, e);
+		}
+		return targetPath;
+	}
+
+	/**
+	 * Retrieve contents from a GitHub repository.
+	 */
+	private Path retrieveGitHubRepositoryContents(GitRepoUrlRef url, Path targetPath) {
+
+		try {
+			URI gitUri = new URI(url.getRepoUrl().toString());
+			String token = this.templateRepositoryProperties.getTokens().get(gitUri.getHost());
+			GitHub github;
+			if (token == null) {
+				github = GitHub.connectAnonymously();
+			}
+			else {
+				github = new GitHubBuilder().withOAuthToken(token).build();
+			}
+			String repo = gitUri.getPath().substring(1);
+			if (repo.endsWith(".git")) {
+				repo = repo.substring(0, repo.length() - 4);
+			}
+			String ref = url.getRef();
+			GHRepository ghRepository = github.getRepository(repo);
+			InputStream inputStream = ghRepository
+					.readZip((inputstream) -> new ByteArrayInputStream(StreamUtils.copyToByteArray(inputstream)), ref);
+			ZipUtil.unpack(inputStream, targetPath.toFile());
```

Comment:
We lose file permissions. Those are coming in fine if I use OS level tools to unzip zip downloaded via gh api. zeroturnaround is essentially super unreliable and thus I never use it. We could switch to using tar.gz files which work better.